File: RoutinUsage/Providers/DeepSeekUsageProvider.swift
import Foundation

struct DeepSeekUsageProvider: UsageProvider {
    static let endpoint = URL(string: "https://api.deepseek.com/user/balance")!
    static let modelsEndpoint = URL(string: "https://api.deepseek.com/models")!

    let descriptor: ProviderDescriptor
    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
        self.descriptor = ProviderRegistry.builtInDescriptors.first(where: { $0.id == .deepseek })!
    }

    func metricCapabilities(for configuration: KeyConfiguration) -> [UsageMetricCapability] {
        guard configuration.credentialKind == .apiKey else { return [] }
        let warningThreshold = configuration.metadata["balanceWarningThreshold"]
            .flatMap { Decimal(string: $0) }
        return [
            UsageMetricCapability(
                metricID: "balance",
                label: "余额",
                presentation: .balance,
                semantic: .balance,
                isMenuBarSelectable: true,
                menuBarPriority: 0,
                defaultAlertEnabled: true,
                defaultAbsoluteAlertThreshold: warningThreshold
            ),
            UsageMetricCapability(
                metricID: "availability",
                label: "账户状态",
                presentation: .status,
                semantic: .status,
                isMenuBarSelectable: false,
                menuBarPriority: nil,
                defaultAlertEnabled: false,
                defaultAbsoluteAlertThreshold: nil
            )
        ]
    }

    func validate(_ credential: ProviderCredential, now: Date) async throws -> UsageSnapshot? {
        try await fetchUsage(credential, now: now)
    }

    func fetchUsage(_ credential: ProviderCredential, now: Date) async throws -> UsageSnapshot? {
        guard credential.providerID == .deepseek, credential.kind == .apiKey, !credential.secret.isEmpty else {
            throw UsageProviderError.invalidCredential
        }

        var request = URLRequest(url: Self.endpoint)
        request.httpMethod = "GET"
        request.timeoutInterval = 15
        request.setValue("Bearer \(credential.secret)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            throw UsageProviderError.transport
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw UsageProviderError.invalidResponse
        }
        guard (200..<300).contains(httpResponse.statusCode) else {
            switch httpResponse.statusCode {
            case 401, 403:
                throw UsageProviderError.unauthorized
            case 429:
                throw UsageProviderError.rateLimited
            case 500...599:
                throw UsageProviderError.providerUnavailable
            default:
                throw UsageProviderError.invalidResponse
            }
        }

        let allowedModels: [String]
        do {
            allowedModels = try await fetchAllowedModels(credential)
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            allowedModels = []
        }

        do {
            let payload = try JSONDecoder().decode(DeepSeekBalanceResponse.self, from: data)
            guard let balance = payload.balanceInfos.first else {
                throw UsageProviderError.invalidResponse
            }
            let currency = balance.currency.isEmpty ? "CNY" : balance.currency
            let total = Decimal(string: balance.totalBalance) ?? 0
            let granted = Decimal(string: balance.grantedBalance) ?? 0
            let toppedUp = Decimal(string: balance.toppedUpBalance) ?? 0
            let warningThreshold: Decimal?
            if let rawThreshold = credential.metadata["balanceWarningThreshold"] {
                warningThreshold = Decimal(string: rawThreshold)
            } else {
                warningThreshold = nil
            }
            let health = UsageMetricHealthEvaluator.balanceState(
                balance: total,
                warningThreshold: warningThreshold,
                isAvailable: payload.isAvailable
            )

            return UsageSnapshot(
                planName: "API 余额",
                kind: .periodic,
                fiveHour: nil,
                weekly: nil,
                token: nil,
                allowedModels: allowedModels,
                fetchedAt: now,
                providerID: .deepseek,
                credentialID: credential.credentialID,
                metrics: [
                    NormalizedUsageMetric(
                        id: "balance",
                        label: "余额",
                        value: total,
                        unit: .currency,
                        presentation: .balance,
                        semantic: .balance,
                        currencyCode: currency,
                        healthState: health
                    ),
                    NormalizedUsageMetric(
                        id: "grantedBalance",
                        label: "赠金余额",
                        value: granted,
                        unit: .currency,
                        presentation: .value,
                        semantic: .value,
                        currencyCode: currency,
                        healthState: health
                    ),
                    NormalizedUsageMetric(
                        id: "toppedUpBalance",
                        label: "充值余额",
                        value: toppedUp,
                        unit: .currency,
                        presentation: .value,
                        semantic: .value,
                        currencyCode: currency,
                        healthState: health
                    ),
                    NormalizedUsageMetric(
                        id: "availability",
                        label: "账户状态",
                        value: payload.isAvailable ? 1 : 0,
                        unit: .boolean,
                        presentation: .status,
                        semantic: .status,
                        healthState: payload.isAvailable ? health : .unavailable
                    )
                ]
            )
        } catch let error as UsageProviderError {
            throw error
        } catch {
            throw UsageProviderError.invalidResponse
        }
    }

    private func fetchAllowedModels(_ credential: ProviderCredential) async throws -> [String] {
        var request = URLRequest(url: Self.modelsEndpoint)
        request.httpMethod = "GET"
        request.timeoutInterval = 15
        request.setValue("Bearer \(credential.secret)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            throw UsageProviderError.transport
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw UsageProviderError.invalidResponse
        }
        guard (200..<300).contains(httpResponse.statusCode) else {
            throw UsageProviderError.invalidResponse
        }
        let payload = try JSONDecoder().decode(DeepSeekModelsResponse.self, from: data)
        return payload.models
            .compactMap(\.id)
            .filter { !$0.isEmpty }
    }
}

private struct DeepSeekBalanceResponse: Decodable {
    let isAvailable: Bool
    let balanceInfos: [DeepSeekBalanceInfo]

    private enum CodingKeys: String, CodingKey {
        case isAvailable = "is_available"
        case balanceInfos = "balance_infos"
    }
}

private struct DeepSeekBalanceInfo: Decodable {
    let currency: String
    let totalBalance: String
    let grantedBalance: String
    let toppedUpBalance: String

    private enum CodingKeys: String, CodingKey {
        case currency
        case totalBalance = "total_balance"
        case grantedBalance = "granted_balance"
        case toppedUpBalance = "topped_up_balance"
    }
}

private struct DeepSeekModelsResponse: Decodable {
    let models: [DeepSeekModelIdentifier]

    private enum CodingKeys: String, CodingKey {
        case models = "data"
    }
}

private struct DeepSeekModelIdentifier: Decodable {
    let id: String?
}

File: android/provider/src/main/kotlin/ai/routin/mytoken/provider/deepseek/DeepSeekUsageProvider.kt
package ai.routin.mytoken.provider.deepseek

import ai.routin.mytoken.domain.model.Credential
import ai.routin.mytoken.domain.model.CredentialKind
import ai.routin.mytoken.domain.model.CredentialSecret
import ai.routin.mytoken.domain.model.ProviderId
import ai.routin.mytoken.domain.model.UsageMetric
import ai.routin.mytoken.domain.model.UsageMetricHealthState
import ai.routin.mytoken.domain.model.UsageMetricPresentation
import ai.routin.mytoken.domain.model.UsageMetricSemantic
import ai.routin.mytoken.domain.model.UsageMetricUnit
import ai.routin.mytoken.domain.model.UsageSnapshot
import ai.routin.mytoken.domain.usage.UsageProvider
import ai.routin.mytoken.domain.usage.UsageProviderException
import ai.routin.mytoken.provider.health.UsageMetricHealthEvaluator
import ai.routin.mytoken.provider.http.HttpTransport
import ai.routin.mytoken.provider.http.ProviderHttpRequest
import ai.routin.mytoken.provider.http.ProviderHttpResponse
import ai.routin.mytoken.provider.json.ProviderJson
import ai.routin.mytoken.provider.json.arrayOrNull
import ai.routin.mytoken.provider.json.asObjectOrNull
import ai.routin.mytoken.provider.json.boolOrNull
import ai.routin.mytoken.provider.json.decimalOrNull
import ai.routin.mytoken.provider.json.objOrNull
import ai.routin.mytoken.provider.json.stringOrNull
import java.math.BigDecimal
import java.time.Clock

/**
 * DeepSeek adapter, aligned with the macOS DeepSeekUsageProvider: GET /user/balance
 * with Bearer auth and a best-effort GET /models request for the allowed model list.
 */
class DeepSeekUsageProvider(
    private val transport: HttpTransport,
    private val clock: Clock = Clock.systemUTC(),
    private val endpoint: String = DEFAULT_ENDPOINT,
) : UsageProvider {

    override val providerId: ProviderId = ProviderId.DeepSeek

    override suspend fun fetchUsage(credential: Credential, secret: CredentialSecret): Result<UsageSnapshot> {
        if (credential.providerId != ProviderId.DeepSeek || credential.credentialKind != CredentialKind.ApiKey) {
            return Result.failure(UsageProviderException.InvalidCredential())
        }
        val key = (secret as? CredentialSecret.ApiKey)?.key
            ?: return Result.failure(UsageProviderException.InvalidCredential())
        if (key.isEmpty()) return Result.failure(UsageProviderException.InvalidCredential())

        val response = try {
            transport.execute(
                ProviderHttpRequest(
                    method = "GET",
                    url = endpoint,
                    headers = mapOf(
                        "Authorization" to "Bearer $key",
                        "Accept" to "application/json"
                    )
                )
            )
        } catch (error: Exception) {
            if (error is kotlinx.coroutines.CancellationException) throw error
            return Result.failure(UsageProviderException.Transport())
        }

        val allowedModels = runCatching {
            val modelsResponse = transport.execute(
                ProviderHttpRequest(
                    method = "GET",
                    url = MODELS_ENDPOINT,
                    headers = mapOf(
                        "Authorization" to "Bearer $key",
                        "Accept" to "application/json"
                    )
                )
            )
            readModelIds(modelsResponse)
        }.getOrElse { error ->
            if (error is kotlinx.coroutines.CancellationException) throw error
            emptyList()
        }

        return runCatching { mapResponse(response, credential, allowedModels) }.recoverCatching { error ->
            if (error is UsageProviderException) throw error
            throw UsageProviderException.InvalidResponse()
        }
    }

    private fun mapResponse(
        response: ProviderHttpResponse,
        credential: Credential,
        allowedModels: List<String>,
    ): UsageSnapshot {
        val body = response.body.decodeToString()
        if (response.statusCode !in 200..299) {
            throw classifyStatus(response.statusCode)
        }
        val root = ProviderJson.parse(body).asObjectOrNull()
            ?: throw UsageProviderException.InvalidResponse()
        val isAvailable = root.boolOrNull("is_available") ?: false
        val balanceInfo = root.arrayOrNull("balance_infos")?.firstOrNull()?.asObjectOrNull()
            ?: throw UsageProviderException.InvalidResponse()

        val currency = balanceInfo.stringOrNull("currency")?.takeIf { it.isNotEmpty() } ?: "CNY"
        val total = balanceInfo.decimalOrNull("total_balance") ?: BigDecimal.ZERO
        val granted = balanceInfo.decimalOrNull("granted_balance") ?: BigDecimal.ZERO
        val toppedUp = balanceInfo.decimalOrNull("topped_up_balance") ?: BigDecimal.ZERO
        val health = UsageMetricHealthEvaluator.balanceState(
            balance = total,
            warningThreshold = null,
            isAvailable = isAvailable
        )

        return UsageSnapshot(
            credentialId = credential.id,
            fetchedAt = clock.instant(),
            planName = "API 余额",
            allowedModels = allowedModels,
            metrics = listOf(
                UsageMetric(
                    id = "balance",
                    label = "余额",
                    value = total,
                    unit = UsageMetricUnit.Currency,
                    presentation = UsageMetricPresentation.Balance,
                    semantic = UsageMetricSemantic.Balance,
                    currencyCode = currency,
                    healthState = health
                ),
                UsageMetric(
                    id = "grantedBalance",
                    label = "赠金余额",
                    value = granted,
                    unit = UsageMetricUnit.Currency,
                    presentation = UsageMetricPresentation.Value,
                    semantic = UsageMetricSemantic.Value,
                    currencyCode = currency,
                    healthState = health
                ),
                UsageMetric(
                    id = "toppedUpBalance",
                    label = "充值余额",
                    value = toppedUp,
                    unit = UsageMetricUnit.Currency,
                    presentation = UsageMetricPresentation.Value,
                    semantic = UsageMetricSemantic.Value,
                    currencyCode = currency,
                    healthState = health
                ),
                UsageMetric(
                    id = "availability",
                    label = "账户状态",
                    value = if (isAvailable) BigDecimal.ONE else BigDecimal.ZERO,
                    unit = UsageMetricUnit.Boolean_,
                    presentation = UsageMetricPresentation.Status,
                    semantic = UsageMetricSemantic.Status,
                    healthState = if (isAvailable) health else UsageMetricHealthState.Unavailable
                )
            )
        )
    }

    private fun readModelIds(response: ProviderHttpResponse): List<String> {
        if (response.statusCode !in 200..299) throw UsageProviderException.InvalidResponse()
        val data = ProviderJson.parse(response.body.decodeToString())
            .asObjectOrNull()
            ?.arrayOrNull("data")
            ?: return emptyList()
        return data.mapNotNull { item ->
            item.asObjectOrNull()?.stringOrNull("id")?.takeIf(String::isNotBlank)
        }
    }

    private fun classifyStatus(statusCode: Int): UsageProviderException = when (statusCode) {
        401, 403 -> UsageProviderException.Unauthorized()
        429 -> UsageProviderException.RateLimited()
        in 500..599 -> UsageProviderException.ProviderUnavailable()
        else -> UsageProviderException.InvalidResponse()
    }

    companion object {
        const val DEFAULT_ENDPOINT = "https://api.deepseek.com/user/balance"
        const val MODELS_ENDPOINT = "https://api.deepseek.com/models"
    }
}
